ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLChatAgentRunner.java

```diff
@@ -324,298 +234,120 @@ private void runReAct(
                 if (finalI % 2 == 0) {
                     MLTaskResponse llmResponse = (MLTaskResponse) output;
                     ModelTensorOutput tmpModelTensorOutput = (ModelTensorOutput) llmResponse.getOutput();
-                    Map<String, ?> dataAsMap = tmpModelTensorOutput.getMlModelOutputs().get(0).getMlModelTensors().get(0).getDataAsMap();
-                    if (dataAsMap.size() == 1 && dataAsMap.containsKey("response")) {
-                        String response = (String) dataAsMap.get("response");
-                        String thoughtResponse = extractModelResponseJson(response);
-                        dataAsMap = gson.fromJson(thoughtResponse, Map.class);
-                    }
-                    String thought = String.valueOf(dataAsMap.get("thought"));
-                    String action = String.valueOf(dataAsMap.get("action"));
-                    String actionInput = parseInputFromLLMReturn(dataAsMap);
-                    String finalAnswer = (String) dataAsMap.get("final_answer");
-                    if (!dataAsMap.containsKey("thought")) {
-                        String response = (String) dataAsMap.get("response");
-                        Pattern pattern = Pattern.compile("```json(.*?)```", Pattern.DOTALL);
-                        Matcher matcher = pattern.matcher(response);
-                        if (matcher.find()) {
-                            String jsonBlock = matcher.group(1);
-                            Map map = gson.fromJson(jsonBlock, Map.class);
-                            thought = String.valueOf(map.get("thought"));
-                            action = String.valueOf(map.get("action"));
-                            actionInput = parseInputFromLLMReturn(map);
-                            finalAnswer = (String) map.get("final_answer");
-                        } else {
-                            finalAnswer = response;
-                        }
-                    }
+                    List<String> llmResponsePatterns = gson.fromJson(parameters.get("llm_response_pattern"), List.class);
+                    Map<String, String> modelOutput = parseLLMOutput(tmpModelTensorOutput, llmResponsePatterns);
+
+                    String thought = String.valueOf(modelOutput.get(THOUGHT));
+                    String action = String.valueOf(modelOutput.get(ACTION));
+                    String actionInput = String.valueOf(modelOutput.get(ACTION_INPUT));
```

Comment:
We don't need String.valueOf any more, as the modelOutput is a Map<String, String> now, not Map<String, Object>.

Reply:
This is for the case when some key not in modelOutput , for example if `ACTION_INPUT` not in `modelOutput` , then `modelOutput.get(ACTION_INPUT)` will be null. To keep compatible with the existing code, I use String.valueOf() to avoid null pointer exception.